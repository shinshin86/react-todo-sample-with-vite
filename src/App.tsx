import { useState } from "react";
import "./App.css";
import TodoItem, { Todo } from "./TodoItem";
import Filter, { FilterType } from "./Filter";

const initialTodos: Todo[] = [
  { id: 1, text: "Task AAA", checked: true },
  { id: 2, text: "Task BBB", checked: false },
  { id: 3, text: "Task CCC", checked: false },
];

const DEFAULT_FILTER_TYPE = "all";

function App() {
  const [text, setText] = useState("");
  const [inputError, setInputError] = useState("");
  const [todoList, setTodoList] = useState<Todo[]>(initialTodos);
  const [filter, setFilter] = useState<FilterType>(DEFAULT_FILTER_TYPE);

  const handleOnSubmit = () => {
    if (!text) {
      setInputError("Please enter your todo");
      return;
    }

    setTodoList(
      todoList.concat({ id: todoList.length + 1, text, checked: false }),
    );

    setText("");
    setInputError("");
  };

  const onCheckTodo = (todoId: number) => {
    const checkTodo = todoList.find(({ id }) => id === todoId);
    if (!checkTodo) throw new Error("Not found check todo");

    checkTodo.checked = !checkTodo.checked;

    const mergedTodoList = todoList.map((todo) => {
      if (todo.id === checkTodo.id) {
        return checkTodo;
      }

      return todo;
    });

    setTodoList(mergedTodoList);
  };

  const isShowTodo = (todo: Todo, filter: FilterType): boolean => {
    return filter === "all" || (filter === "active" && !todo.checked) ||
      filter === "completed" && todo.checked;
  };

  return (
    <div className="container">
      <h1>Your TODO</h1>
      <Filter filter={filter} setFilter={setFilter} />
      <form
        onSubmit={(e) => {
          e.preventDefault();
          handleOnSubmit();
        }}
      >
        <input
          type="text"
          value={text}
          onChange={(e) => setText(e.target.value)}
        />
        {inputError && <p style={{ color: "red" }}>{inputError}</p>}
        <button type="submit">Add</button>
      </form>
      <div style={{ padding: 16 }}>
        {todoList.map((todo, index) => {
          if (isShowTodo(todo, filter)) {
            return (
              <TodoItem todo={todo} onCheckTodo={onCheckTodo} key={index} />
            );
          }
        })}
      </div>
    </div>
  );
}

export default App;
