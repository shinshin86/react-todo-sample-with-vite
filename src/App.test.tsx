import { describe, expect, test } from "vitest";
import "@testing-library/jest-dom";
import App from "./App";
import { render, screen } from "@testing-library/react";

describe("Simple working test", () => {
  test("Should render correctly", () => {
    const { container } = render(<App />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("The title is visible", () => {
    render(<App />);
    expect(screen.getByText("Your TODO")).toBeInTheDocument();
  });
});
