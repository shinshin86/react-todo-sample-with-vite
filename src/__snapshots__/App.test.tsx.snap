// Vitest Snapshot v1

exports[`Simple working test > Should render correctly 1`] = `
<div
  class="container"
>
  <h1>
    Your TODO
  </h1>
  <div
    style="padding: 16px;"
  >
    <h2>
      Choose task filter
    </h2>
    <div>
      <input
        checked=""
        defaultvalue="all"
        id="all"
        name="all"
        type="checkbox"
        value="all"
      />
      <label
        for="all"
      >
        All
      </label>
      <input
        defaultvalue="active"
        id="active"
        name="active"
        type="checkbox"
        value="active"
      />
      <label
        for="Active"
      >
        Active
      </label>
      <input
        defaultvalue="completed"
        id="completed"
        name="completed"
        type="checkbox"
        value="completed"
      />
      <label
        for="completed"
      >
        Completed
      </label>
    </div>
  </div>
  <form>
    <input
      defaultvalue=""
      type="text"
    />
    <button
      type="submit"
    >
      Add
    </button>
  </form>
  <div
    style="padding: 16px;"
  >
    <div>
      <input
        checked=""
        id="1"
        type="checkbox"
      />
      <label
        for="1"
      >
        Task AAA
      </label>
    </div>
    <div>
      <input
        id="2"
        type="checkbox"
      />
      <label
        for="2"
      >
        Task BBB
      </label>
    </div>
    <div>
      <input
        id="3"
        type="checkbox"
      />
      <label
        for="3"
      >
        Task CCC
      </label>
    </div>
  </div>
</div>
`;
